core/src/main/scala/cats/data/AndThen.scala

```diff
@@ -152,6 +152,41 @@ object AndThen extends AndThenInstances0 {
    * to be in danger of triggering a stack-overflow error.
    */
   final private val fusionMaxStackDepth = 127
+
+  /**
+   * If you are going to call this function many times, right associating it
+   * once can be a significant performance improvement for VERY long chains.
+   */
+  def toRightAssociated[A, B](fn: AndThen[A, B]): AndThen[A, B] = {
+    @tailrec
+    def loop[X, Y](beg: AndThen[A, X], middle: AndThen[X, Y], end: AndThen[Y, B], endDone: Boolean): AndThen[A, B] =
+      if (endDone) {
+        // end is right associated
+        middle match {
+          case sm @ Single(_, _) =>
+            val newEnd = Concat(sm, end)
+            beg match {
+              case sb @ Single(_, _)  => Concat(sb, newEnd)
+              case Concat(begA, begB) => loop(begA, begB, newEnd, true)
+            }
+          case Concat(mA, mB) =>
+            // rotate mA onto beg:
+            loop(Concat(beg, mA), mB, end, true)
+        }
+      } else {
+        // we are still right-associating the end
+        end match {
+          case se @ Single(_, _)  => loop(beg, middle, se, true)
+          case Concat(endA, endB) => loop(beg, Concat(middle, endA), endB, false)
+        }
+      }
+
+    fn match {
+      case Concat(Concat(a, b), c)                           => loop(a, b, c, false)
+      case Concat(a, Concat(b, c))                           => loop(a, b, c, false)
+      case Concat(Single(_, _), Single(_, _)) | Single(_, _) => fn
```

Comment:
we could potentially combine adjacent `Single` if the sum of their stack depth is `< fusionMaxStackDepth`? Such a case might have been hidden by a non-right associated construction.